spirv-opt: Eliminate Dead Input Components Pass

greg-lunarg opened the pull request:
This adds the --eliminate-dead-input-components pass which currently
removes trailing unused components from input arrays.

Fixes #4532

dneto0 commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,132 @@
+      continue;
+    }
+    unsigned original_max =
+        arr_len_inst->GetSingleWordInOperand(kConstantValueInIdx) - 1;
```
The array length constant could be a signed integer.  But SPIR-V requires the size to be at least 1, so there is no danger of going negative here.
It might be worth a comment to that effect.

dneto0 commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,132 @@
+        if (seen_non_const_ac) return;
+        auto use_opcode = use->opcode();
+        if (use_opcode == SpvOpLoad || use_opcode == SpvOpCopyMemory ||
+            use_opcode == SpvOpCopyMemorySized) {
```
Please add OpCopyObject to this list, or add bookeeping to also scan the uses of its result ID.


This pass assumes Shader (because of assumptions that we're looking at logical points), and that should be documented and probably checked.

I looked for potential uses of pointers in extended instructions. The InterpolateAt* instructions take a pointer-to-input operand, but only for scalar and vector pointee types. So I think they don't need to be addressed here.

greg-lunarg replied:
OK. For the moment we will just bail if OpCopyObject is seen, leaving the additional bookkeeping for future improvement. Thanks for the due diligence!

greg-lunarg replied:
And for the Shader assumption, I have added a check. I assumed Input implied Shader, but that was wrong. Good catch.

dneto0 commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,132 @@
+  // Move array OpVariable after its type
+  Instruction* new_ptr_ty_inst = def_use_mgr->GetDef(new_ptr_ty_id);
+  arr.RemoveFromList();
+  arr.InsertAfter(new_ptr_ty_inst);
```
I think more bookkeeping is needed?  Update uses of the variable that need to know about the new type:
- OpCopyObject on the variable result ID
- Op*AccessChain with the variable as a base, but with zero indices (yeah, a corner case).

greg-lunarg replied:
Since I have since decided to bail on a use by OpCopyObject, there is no additional type change needed for this case.

Also, I have since added logic to bail on a use by OpAccessChain with no indices (good catch!), so no additional type change is required there.

dneto0 reviewed (commented):
Looking good so far.
This is short and sweet. Thanks for the chance to see WIP.

greg-lunarg commented:
@dneto0 I have addressed all of your concerns so far and have added regression tests and moved this out of WIP status. Please let me know if you have any other concerns or would like anyone else to review.

greg-lunarg commented:
@dneto0 Ping :)

greg-lunarg commented:
@diego Is David available to finish this? If not, could you please have someone finish the review? Thanks!

greg-lunarg commented:
@s-perron @jaebaek Is there someone who can please finish this review?

s-perron commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,144 @@
+  analysis::ConstantManager* const_mgr = context()->get_constant_mgr();
+  analysis::DefUseManager* def_use_mgr = context()->get_def_use_mgr();
+  analysis::Pointer* ptr_type = type_mgr->GetType(arr.type_id())->AsPointer();
+  const analysis::Array* arr_ty = ptr_type->pointee_type()->AsArray();
```
You should assert that `arr_ty` is not null.  Just in case someone decides to change the earlier logic, the will hit the assert and know they need to change this function as well.

greg-lunarg replied:
Ok

s-perron commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,144 @@
+  // Move array OpVariable instruction after its new type to preserve order
+  Instruction* new_ptr_ty_inst = def_use_mgr->GetDef(new_ptr_ty_id);
+  arr.RemoveFromList();
+  arr.InsertAfter(new_ptr_ty_inst);
```
You should assert that the arr is not Function scope (and therefore should not be moved).  I know this will currently be true, but I could see someone updating this to work on all array types for some reason.  A function should be asserting for every assumption that it makes to make it easier to know if the assumptions are wrong of the code that calls this function change.

greg-lunarg replied:
Ok

s-perron commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,144 @@
+          return;
+        }
+        if (use->opcode() != SpvOpAccessChain &&
+            use->opcode() != SpvOpInBoundsAccessChain) {
```
There is no test that uses OpInBoundsAccessChain.  Could you add one?

greg-lunarg replied:
Ok

greg-lunarg replied:
Ok

s-perron commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,144 @@
+        }
+        // OpAccessChain with no indices currently not optimized
+        if (use->NumInOperands() == 1) {
+          seen_non_const_ac = true;
```
No test for this case.  Can you add one?

greg-lunarg replied:
Ok

s-perron commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,144 @@
+        }
+        unsigned base_id = use->GetSingleWordInOperand(kAccessChainBaseInIdx);
+        if (base_id != inst.result_id()) {
+          seen_non_const_ac = true;
```
I don't think this is possible because |inst| is always a variable.  If it was not the base id, then it would be invalid spir-v, right?  If that is correct, I think this should be an assert.

s-perron commented on source/opt/eliminate_dead_input_components_pass.h:
```diff
@@ -0,0 +1,58 @@
+
+ private:
+  // Find the max constant used to index the variable declared by |inst|.
+  // If any non-constant indices, return |original_max|.
```
I would change this description a little.  For example, and `OpLoad` causes it to return `original_max`, but no values was uses to index the variable.  Describing it in terms of the max index that is accessed might be better:

> Returns the index of the element one past the last component that could possibly be accessed.

greg-lunarg replied:
Your new text is not quite accurate, but I have improved the description to be more accurate.

s-perron commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,144 @@
+}
+
+unsigned EliminateDeadInputComponentsPass::FindMaxIndex(Instruction& inst,
+                                                        unsigned original_max) {
```
Add an assert to make explicit the assumptions you make about `inst`.  I think all that is really needed is that `inst` is an `OpVariable`.

greg-lunarg replied:
Ok.

s-perron commented on source/opt/eliminate_dead_input_components_pass.cpp:
```diff
@@ -0,0 +1,144 @@
+                                                        unsigned original_max) {
+  unsigned max = 0;
+  bool seen_non_const_ac = false;
+  context()->get_def_use_mgr()->ForEachUser(
```
I think you can use `WhileEachUser` to stop the search short if you hit something that potentially accesses the last element of the array.

greg-lunarg replied:
Ok.

s-perron reviewed (changes requested):
Generally looks good.  Just a few minor issue with coding style, and a couple extra tests would be nice.  Thanks for this optimization.
